### PR TITLE
Add EnhancedPageObjects definitions

### PR DESCRIPTION
**Pull request by ClaytonAstrom**
### Intro
So this will be a pretty hefty change, but I think we're kidding ourselves by excluding Page Objects.

First and foremost, there's a limited scope of functions the NightwatchAPI and Page Object API share, so I moved those to the SharedFunctions interface.

Secondly, this is not a complete definition.  I'm hoping other people understand some of the other properties better in order to implement them (Sections, url/nagivate, etc.)

Thirdly, the main point of the change is so we can properly use `this` context in Page Objects.  

### Usage

If you want to create a page object, you should have a definition for it in your project -
```ts
export interface TestPage extends EnhancedPageObject {
  foo: () => TestPage;
  bar: (foobar: string) => TestPage;
}

const testPageCommands = {
  foo(this: TestPage) {
    return this.click('@testElement');
  },
  bar(this: TestPage, foobar: string) {
    this.api.elements(foobar, this.elements.testElement.selector, result => {
      console.log(result.length);
    }); // if we want to use the NightwatchAPI elements function, we have to change scope to this.api
    //  Likewise, this.elements returns the map of elements on the current page object
  }
};

module.exports = {
  commands: [testPageCommands],
  elements: {
    testElement: {
      selector: 'div.class'
    }
  }
};
```
And caste your instances -
```ts
module.exports = {
  'Test definitions': (browser: NightwatchBrowser) => {
    {
      const testPage = browser.page.testPage() as TestPage;
      testPage.foo().bar('div.findMe');
    }
  }
};
```

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/nightwatchjs/nightwatch/wiki/Page-Object-API
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by ClaytonAstrom** [comment]
This might need a major version change I think?

**Comment by Schlesiger** [comment]
Perhaps. Is this compatible with the [`nightwatch.js` v1.0 beta](https://www.npmjs.com/package/nightwatch/v/1.0.11)? Otherwise I think it'd be best to wait before bumping the major version.

**Review by wallzero — COMMENTED** [review]
You should consider [adding yourself as a maintainer](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d6b4b59d4928d59e03e841797dad6cb5a22000a9/types/nightwatch/index.d.ts#L3). I am hesitant approving this request because I am at least a couple more months from getting back into the Nightwatch/E2E space and can't assist others if they have questions about these changes.

**Comment by ClaytonAstrom** [comment]
I'm not sure if all the existing definitions are compatible (nothing threw a fit whenever I switched back and forth), but the page objects should not have changed.

The reason I suggest a major version change is this could be potentially breaking definitions, if people have been "working around" typescript:
```ts
foo(this: Nightwatch) {
  return this.api.elements();
}
```
But maybe forcing those breaks is for the best.

Sure, I can update that in the next PR

**Comment by Schlesiger** [comment]
Yeah it's a tough call. As far as I can tell, DefinitelyTyped doesn't support version tags when publishing to NPM; so we can't release an `1.0-alpha`. That being said, it's never a good idea to break semver. After some thought I think you are right; we should bump the major version. If the definition has compatibility issues with the `1.0-beta` releases, they can fixed when they are found.

**Review by wallzero — CHANGES_REQUESTED** [review]
* [bump major version](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d6b4b59d4928d59e03e841797dad6cb5a22000a9/types/nightwatch/index.d.ts#L1)
* [add yourself as a maintainer](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d6b4b59d4928d59e03e841797dad6cb5a22000a9/types/nightwatch/index.d.ts#L3)

**Review by wallzero — APPROVED** [review]

File: types/nightwatch/index.d.ts [hunk 1]
```diff
@@ -1,11 +1,12 @@
-// Type definitions for nightwatch 0.9
+// Type definitions for nightwatch 1.0
```
Comment (rbuckton) [inline types/nightwatch/index.d.ts]: Are these changes to support nightwatch@1.0? From the issue comment thread, that does not seem to be the case.

**Review by rbuckton — CHANGES_REQUESTED** [review]
> Yeah it's a tough call. As far as I can tell, DefinitelyTyped doesn't support version tags when publishing to NPM; so we can't release an `1.0-alpha`. That being said, it's never a good idea to break semver. After some thought I think you are right; we should bump the major version. If the definition has compatibility issues with the `1.0-beta` releases, they can fixed when they are found.

Version numbers specified in the header for definition files do not precisely follow semver. Rather, the _major and minor version numbers_ in the definition file header should match the major and minor version numbers of the _package_ that corresponds to the definition. So, if these definitions are intended to correspond to nightwatch v0.9, then the version number in the header should remain `0.9`. If these definitions are intended to correspond to nightwatch 1.0, then the version number in the header should be `1.0`. The types publisher automatically increments the patch number each time it publishes the definitions to NPM.

Also, when updating a major version, you should follow the steps here: https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-update-a-package-to-a-new-major-version.

**Comment by ClaytonAstrom** [comment]
@Schlesiger - based on the first line `If you intend to continue updating the older version of the package`, I'm changing my argument.  I don't think there's any intent on maintaining two versions unless we started adding `1.x` definitions, and that these changes should be the way forward?  I mean, if anyone is currently using page objects (which they should be) with these definitions, they're definitely "tricking" their way around typescript.  Or they're extending/using their own definitions. 
 Let me know your thoughts

**Comment by Schlesiger** [comment]
@ClaytonAstrom Yes we aren't really targeting `1.0` and the `nightwatch` is still in beta anyhow. Also, these type changes are applicable to `0.9` and would only be a compile-time error, so the breaking change isn't as critical. Let's take the advice and revert back to `0.9`.

@rbuckton Thanks for the help. Is it true we can't target `beta` tags? I couldn't find anything concrete.

**Comment by ClaytonAstrom** [comment]
@rbuckton is there anything else I need to change at this point?

**Comment by ClaytonAstrom** [comment]
@Schlesiger is this PR waiting on anything else that you're aware of?  Not exactly sure what's stuck

**Comment by Schlesiger** [comment]
Looks like we are still waiting on @rbuckton. You could also close this PR and open a new one. I'll approve it right away.

**Comment by ClaytonAstrom** [comment]
Alright, thanks!  Will reopen

**Pull request closed without merging**